winit-core/src/window.rs

```diff
@@ -1118,14 +1124,52 @@ pub trait Window: AsAny + Send + Sync + fmt::Debug {
     ///
     /// [`Ime`]: crate::event::WindowEvent::Ime
     /// [`KeyboardInput`]: crate::event::WindowEvent::KeyboardInput
-    fn set_ime_allowed(&self, allowed: bool);
+    #[deprecated = "use set_ime_state instead"]
+    fn set_ime_allowed(&self, allowed: bool) {
+        let new_state = if allowed { Some(ImeState::new()) } else { None };
+        self.set_ime_state(new_state.as_ref())
+    }
 
     /// Sets the IME purpose for the window using [`ImePurpose`].
     ///
     /// ## Platform-specific
     ///
     /// - **iOS / Android / Web / Windows / X11 / macOS / Orbital:** Unsupported.
-    fn set_ime_purpose(&self, purpose: ImePurpose);
+    #[deprecated = "use set_ime_state instead"]
+    fn set_ime_purpose(&self, purpose: ImePurpose) {
+        if let Some(state) = self.get_ime_state() {
+            let new_state = state.with_purpose(purpose);
+            self.set_ime_state(Some(&new_state));
+        }
+    }
+
+    /// Atomically sets the IME state for the window using [`ImeState`].
+    ///
+    /// If the state is Some, this requests an input method, and the window begins to receive input
+    /// method events.
+    ///
+    /// If the state is None, this disables the input method, and the window stops receiving input
+    /// method events.
+    ///
+    /// Input methods allows the user to compose text without using a keyboard. Requesting one may
+    /// be beneficial for touch screen environments or ones where, for example, East Asian scripts
+    /// may be entered.
+    ///
+    /// If the focus within the application changes from one logical text input area to another, the
+    /// application should inform the IME of the switch by disabling the IME and enabling it again
+    /// in the other area.
+    ///
+    /// IME is **not** enabled by default.
+    ///
+    /// ## Platform-specific
+    ///
+    /// - **iOS / Android / Web / Windows / X11 / macOS / Orbital:** Unsupported.
+    fn set_ime_state(&self, _state: Option<&ImeState>) {}
+
+    /// Returns the current IME state, if the set_ime_state API is supported.
+    fn get_ime_state(&self) -> Option<ImeState> {
+        None
+    }
```

Comment:
Remove this part, no need to fix legacy API for Wayland. And we can fix it internally anyway.

Reply:
If I remove this, I have to remove the legacy API completely because it will stop being able to make set calls, see the other comments.